mobile-browser-based-version/src/helpers/model_definition/model.js

```diff
@@ -0,0 +1,152 @@
+   * If IndexedDB is turned on and the working model exists, then load the
+   * existing model from IndexedDB. Otherwise, create a fresh new one.
+   */
+  async init() {
+    let model;
+    if (this.useIndexedDB) {
+      let modelParams = [
+        this.task.taskId,
+        this.task.trainingInformation.modelId,
+      ];
+      let metadata = await getWorkingModelMetadata(...modelParams);
+
+      if (metadata) {
+        console.log('Loading from memory ');
+        model = await getWorkingModel(...modelParams);
+      } else {
+        console.log('Creating New');
+        model = await this.task.createModel();
+      }
+    } else {
+      console.log('No indexDB');
+      model = await this.task.createModel();
+    }
+
+    this.model = model;
+    this.type =
+      model.getUserDefinedMetadata()['personalizationType'] ??
+      personalizationType.NONE;
+  }
+
+  /**
+   *  Getter for the model.
+   *  @returns the model we want to train.
+   */
+  getModel() {
+    return this.model;
+  }
+
+  /**
+   *  This is a getter for the part of the model we want to share with the network.
+   *  For personalizationType.NONE this is simply the model.
+   *  The behaviour depends on the personalization type of the model.
+   *  @returns the part of the model we want to share.
+   */
+  getSharedModel() {
+    switch (this.type) {
+      case personalizationType.NONE:
+        return this.model;
+
+      case personalizationType.INTEROPERABILITY:
+        return this.model.layers[1];
+    }
+  }
+
+  /**
+   *  Getter for the personalization type.
+   *  @returns the personalization type of the model.
+   */
+  getPersonalizationType() {
+    return this.type;
+  }
+
+  /**
+   * This method updates the type of personalization of the model.
+   *
+   * @param {personalizationType} newType the new type
+   * @returns the model with a new personalization type.
+   */
+  updatePersonalizationType(newType) {
+    switch (newType) {
+      case personalizationType.NONE:
+        this.model = this.getSharedModel();
+        this.model.setUserDefinedMetadata({ personalizationType: newType });
+        this.type = newType;
+        console.log('Updating model to type :' + newType);
+        break;
+
+      case personalizationType.INTEROPERABILITY:
+        if (this.type != newType && this.dataType == 'csv') {
+          this.model = this._createInteroperabilityModel(this.getSharedModel());
+          this.model.setUserDefinedMetadata({ personalizationType: newType });
+          this.type = newType;
+          console.log('Updating model to type :' + newType);
+        } else {
+          console.log('Model already of type :' + newType);
+        }
+        break;
+    }
+  }
+
+  /**
+   * Private method that, given a model, creates an Interoperability Model by wrapping it in between two Interoperability Layers.
+   * @param {Object} model the model we want to wrap
+   * @returns the model wrapped in between two Interoperability Layers.
+   */
```

Comment:
personal model? (and move the implementation into an interop specific file?)

Reply:
Yep, we though about it. It does make sense to have personalized models extend the model wrapper class. 

It just appeared more complicated to switch from one type to the other. I'll think about how to implement that though because it does feel more natural to use class hierarchy :)